Define more granular NFT standard using HIP412 (#565)
hashgraph/hedera-improvement-proposal: HIP/hip-0000-nft-standard.md

The goal is that this HIP defines a granular standard that the entire Hedera NFT community can adopt to set us up for the future.

## Rationale

This token metadata standard has been developed with input from multiple developers and projects in the NFT space. The goal is to provide a more granular NFT standard for "Web3 NFTs", providing a more concrete implementation for HIP-412. 

This HIP focuses on:

1. Providing strict guidelines for artists to define their NFT token metadata
2. Maintaining flexibility via the `properties` object allowing for non-specified fields
3. Maintaining a good level of interoperability and coherence with other standards such as OpenSea
4. Providing sufficient examples for all types of NFTs to find the correct metadata implementation as an artist
  
## Specification

Below is the human-readable schema, presented to maximize clarity. 

```json
{
    "name": "NFT name (required)",
    "creator": "artist (required)",
    "creatorDID": "DID URI (optional)",
    "description": "human readable description of the asset (recommended)",
    "image": "cid or path to the NFT's preview image file (required)",

>     "image_integrity": "SHA-256 digest of the file pointed by the image field (recommended)",

Reviewer: Suggest keeping this the same name as we use in the files array
```suggestion
    "sha256_checksum": "SHA-256 digest of the file pointed by the image field (recommended)",
```

Author: I wanted to make it a bit more specific for the image property but I'm happy to accept this name as well for clarity. The ARC0003 standard defined by Algorand used different naming for the preview image and additional files' checksums.

    "type": "MIME type (required)",
    "format": "standard specification (required - i.e. none, opensea, HIPXXX, hashaxis)",
    "properties": {
        // json object that cover the overarching properties of the token
        "files": [
            {
                "uri": "uri to file (required)",
                "sha256_checksum": "cryptographic hash of the representation of the resource the author expects to load (recommended)",
                "type": "MIME type (required)",
                "is_default_file": "(Type: boolean) indicates if the file is the main file for this NFT (conditionally optional - required when multiple files are listed)",
                "metadata": "metadata object (optional)",
                "metadata_uri": "uri to metadata (optional)",
                "localization": [
                    // optional localization array
                    {
                        "uri": "uri to file (required)",
                        "locale": "language identifier (required)"
                    }

>                 ]

Reviewer: localization should appear at the root of the metadata as well so that the main name, description etc can be localized.

Author: Right, didn't think of that!

            }
        ],
        "collection": "Define your collection name to which the NFT belongs (optional)",
        "external_url": "External URI pointing to an informational page about your collection or specific NFT information i.e. www.mynft/collection/<serial>/ (optional)"
    },
    "attributes": [
        // Can only contain trait types for rarity calculation
        {
            "trait_type": "name of trait (required)",
            "value": "value for this trait (required)",
            "max_value": "maximum possible value for this trait (optional)"
        }
    ]
    // no additional root-level properties allowed - properties array should contain additional properties you may require
}
```

> Below is the field-specific rationale.

Reviewer: It would be useful to include the JSON Schema representation like we did in HIP 10 and 412 so that validators and parsers can make use of it.

For example, a Hedera DID looks like 'did:hedera:mainnet:7Prd74ry1Uct87nZqL3ny7aR7Cg46JamVbJgk8azVgUm;hedera:mainnet:fid=0.0.123'. The DID resolves to a JSON DID Document.

References: 
https://github.com/hashgraph/did-method/blob/master/did-method-specification.md
https://w3c.github.io/did-core/

### description

**Type:** string

**Required**

**Description:** Human-readable description of the NFT.

### image

**Type:** string (CID or URI)

**Required**

**Description:** CID or path to the NFT's **preview image** file. Each NFT should have a preview image set to allow NFT tooling to load a display image quickly. 

For image-based NFTs, you can use a compressed image optimized for the web. For other types of NFTs, you can use any compressed image that gives a visual clue as to what to expect as a user. 

It's recommended to host your file on [IPFS](https://ipfs.io/) and use a service like [Pinata](https://pinata.cloud/) to easily pin your file. Your CID should look like this: `ipfs://<hash>`.

Alternatively, you can use [Arweave](https://www.arweave.org/), receiving a similar CID that looks like this: `ar://<hash>`.

> ### image_integrity

Reviewer: ```suggestion
### sha256_checksum
```

**Type:** string (SHA-256)

**Optional (recommended)**

> **Description:** SHA-256 digest of the file pointed by the `image` property. The `image_integrity` property that contains a cryptographic hash of the representation of the resource the author expects to load. 

Reviewer: ```suggestion
**Description:** SHA-256 digest of the file pointed by the `image` property. The `sha256_checksum` property that contains a cryptographic hash of the representation of the resource the author expects to load. 
```

References: https://w3c.github.io/webappsec-subresource-integrity/

### type

**Type:** string (MIME type)

**Required**

**Description:** Mime formatting shall follow the following format: **type/subtype**. As a rule, mime types are all lowercase. However apps should be programmed to accept any case for robustness.

A list of common mime types can be found here: https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types/Common_types

Note that mime types for directories are not uniformly defined. Some IPFS CIDs point to directories rather than files, so this type is useful. This standard shall define it using the format: text/directory.

### format

**Type:** string

**Required**

**Possible values:** 

- `none`: When not following any standard (not recommneded)
- `opensea`: When following the [OpenSea standard](https://docs.opensea.io/docs/metadata-standards)
- `HIPXXX`: When following a standard defined by HIPs, e.g. `HIP412`

> **Description:** Describes the used standard to display NFT metadata. This property allows for faster parsing by NFT tooling because they don't have to figure out which standard they are dealing with. 

Reviewer: I'd like to propose that we update this to have a reference to a published schema. This would let parsers very quickly identify not only the schema type but also the version of the schema being used over time as the standards evolve. I've added more examples in the main discussion here: https://github.com/hashgraph/hedera-improvement-proposal/discussions/413#discussioncomment-3589586

Author: @justynspooner I've added a JSON schema 👍

---
hip: 0000
title: Define a granular NFT Standard (Extends HIP-412)
author: Michiel Mulders <@michielmulders>
working-group: Michiel Mulders <@michielmulders>, Ashe Oro <@Ashe_Oro>, Hash Axis <@hashaxis>, Rocket May <@rocketmay>, HGP Patches <@HGP_Patches>
type: Informational
needs-council-approval: No

> status: Draft

Reviewer 2: ```suggestion
status: Deferred
```

This token metadata standard has been developed with input from multiple developers and projects in the NFT space. The goal is to provide a more granular NFT standard for "Web3 NFTs", providing a more concrete implementation for HIP-412. 

This HIP focuses on:

1. Providing strict guidelines for artists to define their NFT token metadata
2. Maintaining flexibility via the `properties` object allowing for non-specified fields
3. Maintaining a good level of interoperability and coherence with other standards such as OpenSea
4. Providing sufficient examples for all types of NFTs to find the correct metadata implementation as an artist
  
## Specification

Below is the human-readable schema, presented to maximize clarity. This document also includes a Formal JSON Schema definition to assist in data validation and describe the data in formal terms. **[See the Formal JSON Schema Definition](#formal-json-schema-definition) section below.**

```json
{
	"name": "NFT name (required)",
	"creator": "artist (required)",
	"creatorDID": "DID URI (optional)",
	"description": "human readable description of the asset (required)",
	"image": "cid or path to the NFT's preview image file (required)",

> 	"sha256_checksum": "SHA-256 digest of the file pointed by the image field (recommended)",

Reviewer: Is there a reason we use sha256 over md5? md5 is faster to generate from a file and since we're not using it for storing a password and it's only used as a checksum it should do the job fine. sha256 might be overkill for this.

### format

**Type:** string

**Required**

**Possible values:** 

- `none`: When not following any standard (not recommneded)
- `opensea`: When following the [OpenSea standard](https://docs.opensea.io/docs/metadata-standards)
- `HIPXXX`: When following a standard defined by HIPs, e.g. `HIP412`

**Description:** Describes the used standard to display NFT metadata. This property allows for faster parsing by NFT tooling because they don't have to figure out which standard they are dealing with. 

### properties

**Type:** object

**Required**

**Description:** JSON object that covers the overarching properties of the token. It holds all additional properties the artist wants to define for its NFT. Only the "external_url" property is an optional property you can include.

**It's not allowed to add any non-specified root-level properties to the metadata. Any additional data should go into "properties".**

### properties.external_url

**Type:** string (URI)

**Optional**

**Description:** External URI pointing to an informational page about a collection or specific NFT information i.e. `www.mynft/collection/<serial>/`.

### files

**Type:** array

> **Required**

Reviewer: I don't think this should be required. If the NFT preview image is enough to represent the NFT and there is no need for any higher res imagery, then we shouldn't force them to add any other files here.

Author: See other comment about forcing this. They can always set it to the same image if they really want to have the main image be the preview image. It's not a lot of extra metadata. I prefer forcing this.

**Description:** There are situations (such as mutable metadata) where rather than including the "metadata" object it makes sense to point to a different file. Therefore this is a URI that points to a metadata json file for the file in question. 

**To avoid conflicts, if "metadata" is defined then "metadata_uri" should be ignored.**

An NFT creator has the option of using either "metadata" or "metadata_uri". Metadata should be considered the default behaviour as it minimizes the number of calls that need to be made. Metadata_uri should be used in specific situations where defining the metadata object within the base metadata file is inadequate.

### attributes

**Type:** array

**Required**

**Description:** Required array of attribute objects to calculate the rarity score of NFTs. If you don't wish to implement rarity scores, add an empty array. 

### attributes.trait_type

**Type:** string

**Required**

**Description:** Name of trait.

### attributes.display_type

**Type:** string - Possible display types (but other types are allowed):

- `text` (**default value representation**)
- `percentage` (for integer or number values)
- `boost` (for integer or number values)

> - `datetime` (for string values)

Reviewer: Opensea have a good example of the `date` display_type. We can extend this to support `datetime` as well to include the time. I've added color as well to give some inspiration to how that could work.

https://docs.opensea.io/docs/metadata-standards#date-traits

```suggestion
- `datetime` (for a number which represents the unix timestamp in seconds )
- `date` (for a number which represents the unix timestamp in seconds )
- `color` (for a hexadecimal or rgb string color sequence such as "#00ff44" or "rgb(0,255,0)" )
```

Author: I like this!

		},
		"creator": {
			"type": "string",
			"description": "Identifies the artist name(s)."
		},
		"creatorDID": {
			"type": "string",
			"format": "uri",
			"description": "Points to a decentralized identifier to identify the creator.",
		},
		"description": {
			"type": "string",
			"description": "Describes the asset to which this token represents."
		},
		"image": {
			"type": "string",
			"format": "uri",
			"description": "A URI pointing to a preview image resource."
		},
		"sha256_checksum": {
			"type": "string",
			"description": "Cryptographic hash of the representation of the 'image' resource."
		},
		"type": {
			"type": "string",
			"description": "Sets the MIME type for the 'image' resource."
		},
		"format": {
			"type": "string",
			"description": "Name of the format or schema used by the NFT."
		},
		"properties": {
			"type": "object",
			"properties": {
				"external_url": {
					"type": "string",
					"format": "uri",
					"description": "A URI pointing to an informational page about the NFT."
				}
			},
			"additionalProperties": true
		},
		"files": {
			"type": "array",
			"contains": {
				"type": "object"
			},

> 			"minContains": 1,

Reviewer: As mentioned above, I don't think we need to force a files array if the image is enough for the NFT

Author: I would opt to force this because otherwise we will get a mix between people who use the preview image as the main image and some who don't. If you look at other specifications, they even provide guidelines for maximum dimensions for the preview image? We wanted to be a bit more specific so I don't think this hurts?

						"description": "Cryptographic hash of the representation of the 'uri' resource."
					},
					"type": {
						"type": "string",
						"description": "Sets the MIME type for the 'image' resource."
					},
					"is_default_file": {
						"type": "boolean",
						"description": "Indicates if this file object is the main file representing the NFT."
					},
					"metadata": {
						"type": "object",
						"description": "Represents a nested metadata object for the file."
					},
					"metadata_uri": {
						"type": "string",
						"format": "uri",
						"description": "A URI pointing to a metadata resource."
					},
				},
				"required": [
					"uri",
					"type"
				],
				"additionalProperties": false
			}
		},
		"attributes": {
			"type": "array",
			"items": {
				"type": "object",
				"properties": {
					"trait_type": {
						"type": "string",
						"description": "Name of trait."
					},
					"display_type": {
						"type": "string",
						"description": "Sets the representation of the value of the trait."
					},
					"value": {
						"type": ["string", "integer", "number", "boolean"],
						"description": "Value for trait."
					},
					"max_value": {

> 						"type": "string",

Reviewer: This could be a numerical max too
```suggestion
						"type": ["string", "integer", "number"],
```

						"description": "Maximum value for trait."
					}
				},
				"required": [
					"trait_type",
					"value"
				],
				"additionalProperties": false
			}
		},
		"localization": {
			"type": "array",
			"items": {
				"type": "object",
				"properties": {
					"locale": {
						"type": "string",
						"description": "Sets the two-letter language code for the localization resource."
					},
					"is_default_locale": {
						"type": "boolean",
						"description": "Indicates if the set locale is the default locale for this metadata file."
					},
					"metadata_uri": {
						"type": "string",
						"format": "uri",
						"description": "A URI pointing to a metadata resource."
					}
				},
				"required": [
					"locale"
				],
				"additionalProperties": false
			}
		}
	},
	"required": [

> 		"name",
> 		"creator",
> 		"image",
> 		"type",
> 		"format",
> 		"files",
> 		"properties",
> 		"attributes"

Reviewer: I think these would be the minimum required properties for a very simple NFT. I think we're OK adding image and type as required now if we're pushing for it to be the preview image. I can't remember the react exact reason we made image optional in HIP 412, feel free to chime in if anyone sees a reason why this shouldn't be required.

```suggestion
		"name",
		"image",
		"type"
```

Author: As this will become a reference implementation under HIP412, this should be specific. But you are right about setting `properties` and `attributes` to be optional. The rest should be required.

Reviewer 3: “image” was originally listed as ‘required’, however HashAxis noted that in the case of file NFTs, creating both image and file would require two files to pin instead of one. At the scale of millions, this doubling of the number of pinned images is significantly impactful to the cost of hosting the files. For files that can have their thumbnails generated programmatically, an image is not required. Because of this, the “image” field was changed to Conditionally Optional.
